crates/bitwarden-crypto/src/store/context.rs

```diff
@@ -160,16 +160,41 @@ impl<Ids: KeyIds> KeyStoreContext<'_, Ids> {
     ///
     /// # Arguments
     ///
-    /// * `encryption_key` - The key id used to encrypt the `key_to_encrypt`. It must already exist
+    /// * `wrapping_key` - The key id used to wrap(encrypt) the `key_to_wrap`. It must already exist
     ///   in the context
-    /// * `key_to_encrypt` - The key id to encrypt. It must already exist in the context
+    /// * `key_to_wrap` - The key id to wrap. It must already exist in the context
     pub fn encrypt_symmetric_key_with_symmetric_key(
         &self,
-        encryption_key: Ids::Symmetric,
-        key_to_encrypt: Ids::Symmetric,
+        wrapping_key: Ids::Symmetric,
+        key_to_wrap: Ids::Symmetric,
     ) -> Result<EncString> {
-        let key_to_encrypt = self.get_symmetric_key(key_to_encrypt)?;
-        self.encrypt_data_with_symmetric_key(encryption_key, &key_to_encrypt.to_vec())
+        use SymmetricCryptoKey::*;
+
+        let wrapping_key_instance = self.get_symmetric_key(wrapping_key)?;
+        let key_to_wrap_instance = self.get_symmetric_key(key_to_wrap)?;
+        // Aes256CbcHmacKey can wrap keys by encrypting their byte serialization obtained using
+        // SymmetricCryptoKey::to_encoded(). XChaCha20Poly1305Key need to specify the
+        // content format to be either octet stream, in case the wrapped key is a Aes256CbcHmacKey
+        // or Aes256CbcKey, or by specifying the content format to be CoseKey, in case the
+        // wrapped key is a XChaCha20Poly1305Key.
+        match (wrapping_key_instance, key_to_wrap_instance) {
+            (Aes256CbcHmacKey(_), Aes256CbcHmacKey(_) | Aes256CbcKey(_)) => self
+                .encrypt_data_with_symmetric_key(
+                    wrapping_key,
+                    key_to_wrap_instance.to_encoded().as_slice(),
+                ),
+            (XChaCha20Poly1305Key(_), Aes256CbcHmacKey(_) | Aes256CbcKey(_)) => {
+                // These keys should be represented as octet stream payloads in COSE
+                todo!()
```

Comment:
**issue:** I'm hesitant to accept `todo!` in main since they result in a panic, and we'd probably be better of with an error message.

@dani-garcia thoughts?

Reply:
Hmm.. from our contributing docs:

https://contributing.bitwarden.com/contributing/pull-requests/branching#horizontal-separation

> It is accepted and expected that incomplete code will be merged -- either to main behind a feature flag or into the long-lived feature branch.

In this case, it counts as behind a feature flag, since there is no path to end up with an XChaCha20 key. Getting those keys would only ever be enabled, when all panicing codepaths are removed. KM will need to ensure either way that old, unsupported clients show a proper error message / are prevented from sync for accounts with the new key type, which would include clients that have SDK versions that are released with the above code in it.

(Also note, I was recommended to switch to using a todo here: https://github.com/bitwarden/sdk-internal/pull/181#discussion_r2049494165).

Reply:
A middle ground if we don't want panics but want the operation to fail rather than return invalid encrypted data would be to just return an error. We can add a todo comment to show that the behavior is temporary:

```suggestion
                // TODO(KM-xxxx): Implement this
                return Err(CryptoError::OperationNotSupported(...)) // Or a different error, if we prefer
```

Reply:
I'd prefer the error since feature flags are a "runtime" concept and you should not be able to configure the application to crash itself.

Reply:
Okay, it has been updated to return the suggested OperationNotSupported error.